plugins/pluggable/pluggable.tsx

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+
+import {Theme} from 'mattermost-redux/types/themes';
+
+import {GlobalState} from 'types/store';
+
+type Props = {
+
+    /*
+     * Override the component to be plugged
+     */
+    pluggableName: string;
+
+    /*
+     * Components for overriding provided by plugins
+     */
+    components: GlobalState['plugins']['components'];
+
+    /*
+     * Logged in user's theme
+     */
+    theme: Theme;
+
+    /*
+     * Id of the specific component to be plugged.
+     */
+    pluggableId?: string;
+
+    /*
+     * Name of the sub component to use. Defaults to 'component' if unspecified.
+     */
+    subComponentName?: 'mainComponent' | 'headerCentreComponent' | 'headerRightComponent';
+
+    /*
+     * Accept any other prop to pass onto the plugin component
+     */
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [name: string]: any;
+}
+
+type BaseChildProps = {
+    theme: Theme;
+}
+
+export default function Pluggable(props: Props): JSX.Element | null {
+    const {
+        components,
+        pluggableId,
+        pluggableName,
+        subComponentName = '',
+        theme,
+        ...otherProps
+    } = props;
+
+    if (!pluggableName || !Object.hasOwnProperty.call(components, pluggableName)) {
+        return null;
+    }
+
+    // Override the default component with any registered plugin's component
+    // Select a specific component by pluginId if available
+    let content;
+
+    if (pluggableName === 'Product') {
```

Comment:
Why do we have different behavior here for `Product` now? Can't we just support the subcomponent generically?

Reply:
I had to do that to satisfy the typing because that was the only way for me to tell apart when something is a `ProductComponent` instead of a `PluginComponent`. With the way those types are set up, we can't just access arbitrary fields for arbitrary subcomponents. If data was stored in redux in a different way, that'd be doable, but I don't know what the potential side effects of changing that structure would be.

I'm honestly not 100% sure what the purpose of subcomponents is. It seems to be for letting a plugin register multiple components as a batch which makes sense to me, but the structure of the data in the store ends up being confusing since we're mixing types. I think the structure of `Pluggable` somewhat reflects that, especially in the previous version that also supported completely replacing a component (a feature that we haven't used in a long time).

Reply:
I think I created sub-components as a response to the types being all messed up back there and I couldn't do the solution I wanted. Maybe it would be worth fixing all that. This solution is not ideal because it makes the feature work for only one pluggable, without mentioning that in the docs above.

Reply:
I've tried figuring out how to reintegrate that code path, but I wasn't able to make arbitrary fields for subcomponents work. I think we'll need to restructure the plugin state to do that. Ideally, ProductComponent and PluginComponent would be similar enough then to offer subcomponents for both, but as it stands, I don't think I can get that working right now, especially since we don't have anywher that we want Pluggables to use subcomponents yet.